File: webui/src/main/scala/com/taskadapter/webui/pages/SelectConnectorComponent.scala
package com.taskadapter.webui.pages

import com.taskadapter.PluginManager
import com.taskadapter.webui.Page.message
import com.vaadin.flow.component.button.Button
import com.vaadin.flow.component.html.Label
import com.vaadin.flow.component.orderedlayout.VerticalLayout

import scala.collection.JavaConverters._

class SelectConnectorWizardStep(pluginManager: PluginManager, next: String => Unit) extends WizardStep[String] {
  var result: String = ""

  override def getResult: String = result

  override def ui(config: Any) = new SelectConnectorComponent(pluginManager, (connectorId) => {
    result = connectorId
    next(connectorId)
  }).layout
}

class SelectConnectorComponent(pluginManager: PluginManager, next: String => Unit) {
  val layout = new VerticalLayout()
  layout.setSpacing(true)
  layout.setMargin(true)
  layout.add(new Label(message("newConfig.selectSystem")))

  createSystemList()

  private def createSystemList(): Unit = {
    pluginManager.getPluginDescriptors.asScala.foreach { connector =>

      val systemButton = new Button(connector.label)
      systemButton.setWidth("200px")
      systemButton.addClickListener(_ => next(connector.id))
      layout.add(systemButton)
    }
  }

}

File: webui/src/main/java/com/taskadapter/webui/config/SelectConnectorComponent.java
package com.taskadapter.webui.config;

import com.taskadapter.PluginManager;
import com.taskadapter.webui.Page;
import com.vaadin.flow.component.button.Button;
import com.vaadin.flow.component.html.Label;
import com.vaadin.flow.component.orderedlayout.VerticalLayout;

import java.util.function.Function;

public class SelectConnectorComponent extends VerticalLayout {

    private final PluginManager pluginManager;
    private final Function<String, Void> next;

    public SelectConnectorComponent(PluginManager pluginManager, Function<String, Void> next) {
        this.pluginManager = pluginManager;
        this.next = next;
        setSpacing(true);
        setMargin(true);
        add(new Label(Page.message("newConfig.selectSystem")));

        createSystemList();
    }

    private void createSystemList() {
        pluginManager.getPluginDescriptors().forEachRemaining(connector -> {
            var systemButton = new Button(connector.getLabel(),
                    e -> next.apply(connector.getId()));
            systemButton.setWidth("200px");
            add(systemButton);
        });
    }
}
